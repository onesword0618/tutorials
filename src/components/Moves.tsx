import '../index.css';

import {History} from '../domain/type';
import React from 'react';

type Props = {
  history: History[];
  jumpTo: (move: number) => void;
  currentNumber:number;
}

export const Moves : React.FC<Props> = (props) => {
  const {history,jumpTo,currentNumber} = props;
  return (
    <>
      {history.map((step, move) => {
        const desc = move ? `Go to move #${move}.` : 'Go to game start';
        const location = move ? `(${step.location.col}, ${step.location.row})` : '';
        return (
          <li key={move}>
            <button onClick={() => jumpTo(move)}>
              {move == currentNumber ? <b>{desc}</b> : desc}
              {location}
            </button>
          </li>
        );
      })}
    </>
  );
}
