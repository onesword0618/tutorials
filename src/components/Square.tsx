import '../index.css';

import React from 'react';

type Props = {
  onClick:() => void,
  value: 'X' | 'O'| null,
}

export const Square: React.FC<Props> = (props) => {
  const {onClick,value} = props;
  return (
    <button
      className="square"
      onClick={onClick}
    >
      {value}
    </button>
  );
}
