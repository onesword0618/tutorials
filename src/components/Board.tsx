import '../index.css';

import React from 'react';
import { Square } from '../components/Square';

type Props = {
  onClick:(i:number) => void,
  squares:('X' | 'O'| null)[],
  line:(number[] | null),
};

export const Board : React.FC<Props> = (props) => {
  const {onClick,squares,line} = props;
  function renderSquare(i:number,highlight:(boolean|null)) {
    return (
      <Square
        key={i}
        onClick={() => onClick(i)}
        value={squares[i]}
        highlight={highlight}
      />
    );
  };

  const rows = Array(3);
  for(let i of [0,1,2]) {

    const columns = Array(3);
    for (let j of [0,1,2]) {
      const index = i*3 + j;
      const highlight = line && line.includes(index);
      columns[j] = renderSquare(index,highlight);
    };

    rows[i] = (
      <div key={i} className='board-row'>
        {columns}
      </div>
    );

  };

    return (
      <div>
        {rows}
      </div>
    );
};
