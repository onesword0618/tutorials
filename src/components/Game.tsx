import '../index.css';

import React, { useState } from 'react';

import { Board } from './Board';
import { Moves } from './Moves';
import { findWinner } from '../domain/findWinner';

export const Game : React.FC = () => {
  const [history, setHistory] = useState(
    [{
      squares: Array(9).fill(null),
      location: {
        col: 0,
        row: 0
      }
    }]
  );

  const [stepNumber, setStepNumber] = useState<number>(0);
  const [xIsNext, setXIsNext] = useState<boolean>(true);
  const [toggle, setToggle] = useState<boolean>(true);

  const handleClick = (i: number) => {
    const _history = history.slice(0, stepNumber + 1);
    const current = _history[_history.length - 1];
    const squares = current.squares.slice();

    if (findWinner(squares) || squares[i]) {
      return;
    };

    squares[i] = xIsNext ? 'X' : 'O';
    setHistory(_history.concat([
      {
        squares: squares,
        location: {
          col: i % 3 + 1,
          row: Math.trunc(i / 3 + 1),
        }
      }
    ]));
    setStepNumber(_history.length);
    setXIsNext(!xIsNext);
  };

  const jumpTo = (step: number) => {
    setStepNumber(step);
    setXIsNext(step % 2 === 0);
  };

  const current = history[stepNumber];
  const winner:(string|null) = findWinner(current.squares);

  let status:string;
  if (winner) {
    status = 'Winner: ' + winner;
  } else {
    status = 'Next player: ' + (xIsNext ? 'X' : 'O');
  };

  return (
    <div className="game">
      <div className="game-board">
        <Board
          squares={current.squares}
          onClick={(i) => handleClick(i)}
        />
      </div>
      <div className="game-info">
        <div>{status}</div>
        <div>
          <button onClick={() => {setToggle(!toggle)}}>
            {toggle ? 'Sort descending' : 'Sort ascending'}
          </button>
        </div>
        {
          toggle ?
          <ol>
            <Moves
              history={history}
              jumpTo={jumpTo}
              currentNumber={stepNumber}
            />
          </ol> :
          <ol reversed>
            <Moves
              history={history.reverse()}
              jumpTo={jumpTo}
              currentNumber={stepNumber}
            />
          </ol>
        }
      </div>
    </div>
  );
};
