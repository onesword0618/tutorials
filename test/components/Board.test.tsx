import { Board } from "../../src/components/Board";
import React from "react";
import renderer from "react-test-renderer";

describe("盤面を表現しているコンポーネントの検証", () => {
  describe("盤面が描画されること", () => {
    test("SnapShotと同一であること", () => {
      const squeres: ("X" | "O" | null)[] = [
        "X",
        null,
        null,
        "X",
        null,
        null,
        "X",
        null,
        null,
      ];
      const sut = renderer.create(
        <Board squares={squeres} onClick={() => 1} line={[1, 2, 3]} />
      );

      let target = sut.toJSON();
      expect(target).toMatchSnapshot();
    });
  });
});
