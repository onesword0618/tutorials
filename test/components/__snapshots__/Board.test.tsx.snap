// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`盤面を表現しているコンポーネントの検証 盤面が描画されること SnapShotと同一であること 1`] = `
<div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
</div>
`;
